hub/core/transform/transform.py

```diff
@@ -0,0 +1,166 @@
+from hub.core.compute.provider import ComputeProvider
+from hub.util.exceptions import (
+    InvalidInputDataException,
+    TensorMismatchException,
+    UnsupportedSchedulerException,
+)
+from hub.core.compute import ThreadProvider, ProcessProvider
+from hub.util.transform import merge_index_metas, merge_tensor_metas, transform_sample
+from hub.core.meta.tensor_meta import TensorMeta
+from hub.util.remove_cache import remove_all_cache
+from hub.core.storage.lru_cache import LRUCache
+from hub.core.storage.memory import MemoryProvider
+from hub.core.meta.index_meta import IndexMeta
+from hub.core.tensor import append_tensor
+from hub.api.dataset import Dataset
+from hub.constants import MB
+from typing import Callable, Dict, List, Optional, Tuple
+from itertools import repeat
+import math
+
+# TODO Ensure that all outputs have the same schema
+
+
+def transform(
+    data_in,
```

Comment:
```suggestion
    source_dataset: Dataset,
```

please use static typing for ideally all methods

Reply:
actually, this will cause a circular dependency. then you should only interact with the core. (just take in `key` and `storage` and only call core methods)

Reply:
this doesn't have a restriction. It can be anything, a list, tuple, etc

Reply:
then can you add static typing here for all supported types?